File: NFCommunicator/nfcommunicator/src/main/java/com/neofect/communicator/usb/UsbConnection.java
package com.neofect.communicator.usb;

import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbDeviceConnection;
import android.hardware.usb.UsbEndpoint;
import android.hardware.usb.UsbManager;
import android.hardware.usb.UsbRequest;
import android.os.Build;
import android.util.Log;

import com.neofect.communicator.CommunicationController;
import com.neofect.communicator.Connection;
import com.neofect.communicator.ConnectionType;
import com.neofect.communicator.Device;

import java.nio.ByteBuffer;

/**
 * @author neo.kim@neofect.com
 * @date Nov 16, 2016
 */
public class UsbConnection extends Connection {

	private static final String LOG_TAG = UsbConnection.class.getSimpleName();

	private static final String ACTION_USB_PERMISSION = "com.neofect.communicator.USB_PERMISSION";

	private static final int WRITE_TIMEOUT_MILLIS = 200;
	private static final int READ_BUFFER_SIZE = 16 * 1024;
	private static final int WRITE_BUFFER_SIZE = 16 * 1024;

	private final Object mWriteBufferLock = new Object();
	private byte[] mWriteBuffer = new byte[WRITE_BUFFER_SIZE];

	private Context context;
	private UsbDevice device;
	private UsbDeviceConnection deviceConnection;
	private UsbEndpoint readEndpoint;
	private UsbEndpoint writeEndpoint;
	private UsbSerialDriver driver;
	private Thread readThread;
	private BroadcastReceiver usbEventReceiver;

	public UsbConnection(Context context, UsbDevice device, CommunicationController<? extends Device> controller) {
		super(ConnectionType.USB_SERIAL, controller);
		this.context = context.getApplicationContext();
		this.device = device;
	}

	@Override
	public String getDeviceName() {
		if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
			return device.getProductName();
		} else {
			return device.getDeviceName();
		}
	}

	@Override
	public String getRemoteAddress() {
		return device.getDeviceName();
	}

	@Override
	public String getDescription() {
		return getDeviceName() + "(" + getRemoteAddress() + ")";
	}

	@Override
	public void connect() {
		Log.d(LOG_TAG, "connect:");
		if (deviceConnection != null) {
			Log.e(LOG_TAG, "connect: Already connected!");
			return;
		} else if (device.getInterfaceCount() == 0) {
			// 2016.12.17 neo.kim@neofect.com
			// If you got here and you don't know why, the following link might be helpful.
			// https://code.google.com/p/android/issues/detail?id=159529
			throw new UsbDeviceHasNoInterfaceException("The USB device(" + device.getDeviceName() + ") has no USB interface!");
		}
		registerReceiver();

		final UsbManager usbManager = (UsbManager) context.getSystemService(Context.USB_SERVICE);
		if (usbManager.hasPermission(device)) {
			startConnecting();
		} else {
			// Ask permission for USB connection
			PendingIntent intent = PendingIntent.getBroadcast(context, 0, new Intent(ACTION_USB_PERMISSION), 0);
			Log.d(LOG_TAG, "Requesting permission for USB device '" + device.getDeviceName() + "'...");
			usbManager.requestPermission(device, intent);
		}
	}

	@Override
	public void disconnect() {
		Log.d(LOG_TAG, "disconnect:");
		if (deviceConnection == null) {
			Log.e(LOG_TAG, "disconnect: Already disconnected");
			return;
		}
		cleanUp();
		handleDisconnected();
	}

	private void cleanUp() {
		String message = "cleanUp:";
		if (device != null) {
			message += " device=" + getDescription();
		}
		Log.d(LOG_TAG, message);
		if (readThread != null) {
			readThread.interrupt();
			readThread = null;
		}

		if (usbEventReceiver != null) {
			synchronized (usbEventReceiver) {
				try {
					context.unregisterReceiver(usbEventReceiver);
					Log.d(LOG_TAG, "cleanUp: Receiver unregistered.");
				} catch (IllegalArgumentException e) {
					Log.e(LOG_TAG, "cleanUp: Failed to unregister the receiver!", e);
				}
				usbEventReceiver = null;
			}
		}

		try {
			driver.close();
		} catch (Exception e) {
			Log.e(LOG_TAG, "", e);
		} finally {
			driver = null;
			deviceConnection = null;
		}
	}

	private void startConnecting() {
		handleConnecting();
		try {
			UsbManager usbManager = (UsbManager) context.getSystemService(Context.USB_SERVICE);
			deviceConnection = usbManager.openDevice(device);

			driver = UsbSerialDriverFactory.createDriver(device, deviceConnection);
			UsbEndpoint[] endpoints = driver.open();
			readEndpoint = endpoints[0];
			writeEndpoint = endpoints[1];

			driver.setParameters(115200, 8, UsbSerialDriver.STOPBITS_1, UsbSerialDriver.PARITY_NONE);
		} catch (Exception e) {
			Log.e(LOG_TAG, "Failed to connect to USB device(" + device.getDeviceName() + "). cause=" + e.getMessage());
			cleanUp();
			handleFailedToConnect(e);
			return;
		}

		// Start a thread for reading
		readThread = new UsbReadThread();
		readThread.start();

		Log.i(LOG_TAG, "USB device is connected! connection=" + getDescription());
		handleConnected();
	}

	private synchronized void registerReceiver() {
		if (usbEventReceiver != null) {
			Log.e(LOG_TAG, "USB event receiver is already registered!");
			return;
		}
		IntentFilter filter = new IntentFilter();
		filter.addAction(ACTION_USB_PERMISSION);
		filter.addAction(UsbManager.ACTION_USB_DEVICE_DETACHED);
		usbEventReceiver = createReceiver();
		context.registerReceiver(usbEventReceiver, filter);
		Log.d(LOG_TAG, "USB event receiver is registered.");
	}

	private BroadcastReceiver createReceiver() {
		return new BroadcastReceiver() {
			@Override
			public void onReceive(Context context, Intent intent) {
				UsbDevice device = intent.getParcelableExtra(UsbManager.EXTRA_DEVICE);
				if (device == null || !UsbConnection.this.device.equals(device)) {
					return;
				}
				String action = intent.getAction();
				Log.d(LOG_TAG, "USB event is received. action=" + action + ", device=" + UsbConnection.this.getDescription());

				if (UsbManager.ACTION_USB_DEVICE_DETACHED.equals(action)) {
					Log.i(LOG_TAG, "USB device is detached. device=" + device);
					disconnect();
				} else if (ACTION_USB_PERMISSION.equals(action)) {
					if (intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false)) {
						Log.d(LOG_TAG, "Permission granted for the device " + device);
						startConnecting();
					}  else {
						Log.d(LOG_TAG, "Permission denied for the device " + device);
					}
				}
			}
		};
	}

	@Override
	public void write(byte[] src) {
		if (!isConnected()) {
			Log.e(LOG_TAG, "write: Not connected!");
			return;
		}

		try {
			int offset = 0;
			while (offset < src.length) {
				final int writeLength;
				final int numberOfWrittenBytes;

				synchronized (mWriteBufferLock) {
					final byte[] writeBuffer;

					writeLength = Math.min(src.length - offset, mWriteBuffer.length);
					if (offset == 0) {
						writeBuffer = src;
					} else {
						// bulkTransfer does not support offsets, make a copy.
						System.arraycopy(src, offset, mWriteBuffer, 0, writeLength);
						writeBuffer = mWriteBuffer;
					}

					numberOfWrittenBytes = deviceConnection.bulkTransfer(writeEndpoint, writeBuffer, writeLength, WRITE_TIMEOUT_MILLIS);
				}
				if (numberOfWrittenBytes <= 0) {
					Log.e(LOG_TAG, "Error writing " + writeLength + " bytes at offset " + offset + " length=" + src.length);
					return;
				}
				Log.d(LOG_TAG, "UsbConnection: write: numberOfWrittenBytes=" + numberOfWrittenBytes + " attempted=" + writeLength);
				offset += numberOfWrittenBytes;
			}
		} catch (Exception e) {
			Log.e(LOG_TAG, "write()", e);
		}
	}

	private class UsbReadThread extends Thread {
		@Override
		public void run() {
			byte[] buffer = new byte[READ_BUFFER_SIZE];
			final UsbRequest request = new UsbRequest();
			request.initialize(deviceConnection, readEndpoint);
			try {
				while (UsbConnection.this.isConnected()) {
					ByteBuffer buf = ByteBuffer.wrap(buffer);
					if (!request.queue(buf, buffer.length)) {
						Log.e(LOG_TAG, "UsbReadThread: run: Failed queueing request!");
						disconnect();
						break;
					}

					final UsbRequest response = deviceConnection.requestWait();
					if (response == null) {
						Log.e(LOG_TAG, "UsbReadThread: run: requestWait() returned null!");
						disconnect();
						break;
					}

					int numberOfBytes = buf.position();
					if (numberOfBytes > 0) {
						byte[] readData = new byte[numberOfBytes];
						System.arraycopy(buffer, 0, readData, 0, numberOfBytes);
						handleReadData(readData);
					}
				}
			} catch (Exception e) {
				Log.e(LOG_TAG, "UsbReadThread: run:", e);
				disconnect();
			}
			request.close();
		}
	}

}

File: communicator/src/main/java/com/neofect/communicator/usb/UsbConnection.kt
package com.neofect.communicator.usb

import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.hardware.usb.UsbDevice
import android.hardware.usb.UsbManager
import android.os.Build
import android.util.Log
import com.hoho.android.usbserial.driver.CdcAcmSerialDriver
import com.hoho.android.usbserial.driver.UsbSerialPort
import com.hoho.android.usbserial.driver.UsbSerialProber
import com.neofect.communicator.Connection
import com.neofect.communicator.ConnectionType
import com.neofect.communicator.Controller
import com.neofect.communicator.Device
import com.neofect.communicator.util.ByteRingBuffer
import java.io.IOException
import java.nio.ByteBuffer

/**
 * Created by jhchoi on 2022/07/08
 * jhchoi@neofect.com
 */
class UsbConnection(
    private val context: Context,
    private val device: UsbDevice,
    controller: Controller<out Device>?
) : Connection(ConnectionType.USB_SERIAL, controller) {
    private var usbSerialPort: UsbSerialPort? = null
    private var usbEventReceiver: BroadcastReceiver? = null
    private var usbIoManager: SimpleSerialIoManager? = null

    private var readDataHandlerThread: Thread? = null

    private val readDataCache = ByteRingBuffer(
        2 * 1024 * 1024 // Maximum 2MB
    )
    private val cacheLock = Object()

    override fun connect() {
        registerReceiver()

        val usbManager = context.getSystemService(Context.USB_SERVICE) as UsbManager
        //permission check.
        if (!usbManager.hasPermission(device)) {
            // Ask permission for USB connection
            val intent = PendingIntent.getBroadcast(context, 0, Intent(ACTION_USB_PERMISSION), 0)
            Log.d(LOG_TAG, "Requesting permission for USB device '${device.deviceName}'...")
            usbManager.requestPermission(device, intent)
            return
        }
        startConnecting()
    }

    override fun disconnect() {
        if (usbSerialPort == null) {
            Log.e(LOG_TAG, "disconnect: Already disconnected")
            return
        }
        cleanUp()
        handleDisconnected()
    }

    private fun cleanUp() {

        readDataHandlerThread?.interrupt()
        readDataHandlerThread = null

        usbIoManager?.let {
            it.listener = null
            it.stop()
        }
        usbIoManager = null

        if (usbEventReceiver != null) {
            synchronized(this) {
                try {
                    context.unregisterReceiver(usbEventReceiver)
                    Log.d(LOG_TAG, "cleanUp: Receiver unregistered.")
                } catch (e: IllegalArgumentException) {
                    Log.e(LOG_TAG, "cleanUp: Failed to unregister the receiver!", e)
                }
                usbEventReceiver = null
            }
        }

        runCatching {
            usbSerialPort?.close()
        }
        usbSerialPort = null
    }

    private fun startConnecting() {
        handleConnecting()
        val usbManager = context.getSystemService(Context.USB_SERVICE) as UsbManager
        val usbConnection = usbManager.openDevice(device)
        val driver = UsbSerialProber.getDefaultProber().probeDevice(device) ?: run {
            UsbCustomProber.getCustomProber().probeDevice(device)
        } ?: run {
            throw IllegalStateException("Unknown usb device. driver not found.")
        }

        val usbSerialPort = driver.ports.first()
        runCatching {
            usbSerialPort.open(usbConnection)
            when (usbSerialPort) {
                is CdcAcmSerialDriver.CdcAcmSerialPort -> {
                    usbSerialPort.setParameters(
                        115200,
                        8,
                        UsbSerialPort.STOPBITS_1,
                        UsbSerialPort.PARITY_NONE
                    )
                }
                else -> {
                    usbSerialPort.setParameters(
                        115200,
                        8,
                        UsbSerialPort.STOPBITS_1,
                        UsbSerialPort.PARITY_NONE
                    )
                }
            }

            readDataHandlerThread?.interrupt()

            usbIoManager =
                SimpleSerialIoManager(usbSerialPort, READ_BUFFER_SIZE, inputOutputManagerListener)
            usbIoManager?.start()
            this@UsbConnection.usbSerialPort = usbSerialPort

            handleConnected()

            readDataHandlerThread = ReadDataHandlerThread()
            readDataHandlerThread?.start()


        }.onFailure {
            cleanUp()
            val exception = if (it is Exception) it else IOException("connect failed.")
            handleFailedToConnect(exception)
        }
    }

    @Synchronized
    private fun registerReceiver() {
        if (usbEventReceiver != null) {
            Log.e(LOG_TAG, "USB event receiver is already registered!")
            return
        }
        val filter = IntentFilter()
        filter.addAction(ACTION_USB_PERMISSION)
        filter.addAction(UsbManager.ACTION_USB_DEVICE_DETACHED)
        usbEventReceiver = createReceiver()
        context.registerReceiver(usbEventReceiver, filter)
        Log.d(LOG_TAG, "USB event receiver is registered.")
    }

    private fun createReceiver(): BroadcastReceiver {
        return object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                val device = intent.getParcelableExtra<UsbDevice>(UsbManager.EXTRA_DEVICE)?.takeIf {
                    this@UsbConnection.device == device
                } ?: return
                val action = intent.action
                Log.d(
                    LOG_TAG,
                    "USB event is received. action=$action, device=${this@UsbConnection.description}"
                )
                when (action) {
                    UsbManager.ACTION_USB_DEVICE_DETACHED -> {
                        Log.i(LOG_TAG, "USB device is detached. device=$device")
                        disconnect()
                    }
                    ACTION_USB_PERMISSION -> {
                        if (intent.getBooleanExtra(UsbManager.EXTRA_PERMISSION_GRANTED, false)) {
                            Log.i(LOG_TAG, "Permission granted for the device $device")
                            startConnecting()
                        } else {
                            Log.i(LOG_TAG, "Permission denied for the device $device")
                            cleanUp()
                            handleFailedToConnect(SecurityException("User denied to grant USB permission!"))
                        }
                    }
                }
            }
        }
    }

    override fun write(data: ByteArray?) {
        if (!isConnected) {
            Log.e(LOG_TAG, "write: Not connected!")
            return
        }
        runCatching {
            usbSerialPort?.write(data, WRITE_TIMEOUT_MILLIS)
        }.onFailure { e ->
            Log.e(LOG_TAG, "write()", e)
        }
    }

    override fun getDeviceIdentifier(): String = device.deviceName

    override fun getDeviceName(): String =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            device.productName
        } else {
            device.deviceName
        }

    override fun getDescription(): String = "$deviceName($deviceIdentifier)"

    private inner class ReadDataHandlerThread : Thread() {
        override fun run() {
            Log.d(LOG_TAG, "ReadDataHandlerThread start.")
            runCatching {
                android.os.Process.setThreadPriority(android.os.Process.THREAD_PRIORITY_URGENT_AUDIO)
                var bufferCache = ByteBuffer.allocate(2 * 1024 * 1024).array()
                while (isConnected) {
                    val contentSize = synchronized(cacheLock) {
                        val contentSize = readDataCache.contentSize
                        if (contentSize > bufferCache.size) {
                            Log.d(LOG_TAG, "handle thread buffer alloc. size: $contentSize")
                            bufferCache = ByteBuffer.allocate(contentSize).array()
                        }
                        readDataCache.read(bufferCache, contentSize)
                        contentSize
                    }
                    if (bufferCache.isNotEmpty()) {
                        handleReadData(bufferCache, contentSize)
                    }
                    sleep(1)
                }
            }.onFailure {
                it.printStackTrace()
            }
            Log.d(LOG_TAG, "ReadDataHandlerThread finish.")
        }
    }

    private val inputOutputManagerListener = object : SimpleSerialIoManager.Listener {
        override fun onNewData(data: ByteArray, length: Int) {
            if (data != null) {
                synchronized(cacheLock) {
                    readDataCache.put(data, length)
                }
            }
        }

        override fun onRunError(e: Exception?) {
            Log.e(LOG_TAG, "onRunError. ", e)
            disconnect()
        }

    }

    companion object {
        private val LOG_TAG = UsbConnection::class.java.simpleName
        private val ACTION_USB_PERMISSION = "com.neofect.communicator.USB_PERMISSION"
        private const val WRITE_TIMEOUT_MILLIS = 200
        private const val READ_BUFFER_SIZE = 100
    }
}